File: backend/src/services/styleAnalysisService.js
/**
 * Style Analysis Service
 * Business logic for AI style analysis using Gemini
 */

const geminiService = require('./geminiService');

class StyleAnalysisService {
  /**
   * Analyze clothing image
   * @param {Buffer} imageBuffer - Image buffer
   * @param {string} mimeType - Image MIME type
   * @param {Object} imageInfo - Original image info
   * @returns {Promise<Object>} Analysis result with image info
   */
  async analyzeImage(imageBuffer, mimeType, imageInfo) {
    try {
      // Call Gemini AI for analysis
      const analysis = await geminiService.analyzeClothingImage(imageBuffer, mimeType);

      // Return structured response
      return {
        analysis,
        imageInfo: {
          originalName: imageInfo.originalName,
          size: imageInfo.size,
          mimeType: imageInfo.mimeType
        }
      };
    } catch (error) {
      console.error('Error in analyzeImage:', error);
      throw new Error(`Failed to analyze image: ${error.message}`);
    }
  }

  /**
   * Generate comprehensive style profile from wardrobe
   * @param {Array} wardrobeItems - Array of wardrobe items with analysis
   * @returns {Promise<Object>} Style profile
   */
  async generateStyleProfile(wardrobeItems) {
    try {
      if (!wardrobeItems || wardrobeItems.length === 0) {
        throw new Error('Wardrobe items are required');
      }

      if (wardrobeItems.length < 3) {
        throw new Error('At least 3 items are required to generate a style profile');
      }

      // Extract analysis from items
      const items = wardrobeItems.map(item => item.analysis || item);

      // Generate profile using Gemini
      const profile = await geminiService.generateStyleProfile(items);

      // Enhance profile with additional insights
      const enhancedProfile = this._enhanceProfile(profile, items);

      return enhancedProfile;
    } catch (error) {
      console.error('Error in generateStyleProfile:', error);
      throw new Error(`Failed to generate style profile: ${error.message}`);
    }
  }

  /**
   * Get shopping recommendations based on item analysis
   * @param {Object} itemAnalysis - Analysis of a clothing item
   * @returns {Promise<Object>} Recommendations
   */
  async getRecommendations(itemAnalysis) {
    try {
      if (!itemAnalysis) {
        throw new Error('Item analysis is required');
      }

      // Get recommendations from Gemini
      const recommendations = await geminiService.findSimilarItems(itemAnalysis);

      // Enhance recommendations with additional data
      const enhancedRecommendations = this._enhanceRecommendations(recommendations, itemAnalysis);

      return enhancedRecommendations;
    } catch (error) {
      console.error('Error in getRecommendations:', error);
      throw new Error(`Failed to get recommendations: ${error.message}`);
    }
  }

  /**
   * Enhance profile with additional insights
   * @private
   * @param {Object} profile - Base profile from AI
   * @param {Array} items - Wardrobe items
   * @returns {Object} Enhanced profile
   */
  _enhanceProfile(profile, items) {
    // Calculate additional statistics
    const formalities = items.map(item => item.formality).filter(Boolean);
    const seasons = items.flatMap(item => item.season || []);
    
    return {
      ...profile,
      statistics: {
        totalItems: items.length,
        uniqueTypes: [...new Set(items.map(item => item.itemType))].length,
        mostCommonFormality: this._getMostCommon(formalities),
        seasonalCoverage: [...new Set(seasons)]
      },
      generatedAt: new Date().toISOString()
    };
  }

  /**
   * Enhance recommendations with additional data
   * @private
   * @param {Object} recommendations - Base recommendations from AI
   * @param {Object} itemAnalysis - Original item analysis
   * @returns {Object} Enhanced recommendations
   */
  _enhanceRecommendations(recommendations, itemAnalysis) {
    return {
      ...recommendations,
      originalItem: {
        type: itemAnalysis.itemType,
        style: itemAnalysis.style,
        colors: itemAnalysis.colors
      },
      generatedAt: new Date().toISOString()
    };
  }

  /**
   * Get most common value from array
   * @private
   * @param {Array} arr - Array of values
   * @returns {*} Most common value
   */
  _getMostCommon(arr) {
    if (arr.length === 0) return null;
    
    const counts = {};
    arr.forEach(item => {
      counts[item] = (counts[item] || 0) + 1;
    });

    return Object.keys(counts).reduce((a, b) => counts[a] > counts[b] ? a : b);
  }
}

module.exports = new StyleAnalysisService();

File: backend-python/app/services/style_analysis_service.py
"""
Style Analysis Service
Business logic for AI-powered style analysis
"""
from typing import Dict, List, Any
from services.gemini_service import gemini_service
from services.wardrobe_service import wardrobe_service
from datetime import datetime

class StyleAnalysisService:
    """Service for style analysis operations"""
    
    def analyze_image(self, image_data: bytes, mime_type: str, image_info: Dict) -> Dict[str, Any]:
        """
        Analyze a clothing image and return enhanced results
        
        Args:
            image_data: Image binary data
            mime_type: Image MIME type
            image_info: Image metadata (filename, size, mimetype)
            
        Returns:
            Dict containing analysis results with enhancements
        """
        try:
            # Get AI analysis from Gemini
            analysis = gemini_service.analyze_clothing_image(image_data, mime_type)
            
            # Add metadata
            result = {
                'analysis': analysis,
                'imageInfo': image_info,
                'analyzedAt': datetime.now().isoformat()
            }
            
            return result
            
        except Exception as e:
            raise ValueError(f'Failed to analyze image: {str(e)}')
    
    def generate_style_profile(self, user_id: str) -> Dict[str, Any]:
        print(f'[DEBUG] style_analysis_service.generate_style_profile called for user_id={user_id}')
        """
        Generate a comprehensive style profile based on wardrobe
        
        Returns:
            Dict containing style profile and statistics
        """
        try:
            # Get all wardrobe items for user
            wardrobe_items = wardrobe_service.get_all_items(user_id)
            print(f'[DEBUG] wardrobe_items (user_id={user_id}): {wardrobe_items}')

            # Require minimum items for profile
            if len(wardrobe_items) < 3:
                raise ValueError('Need at least 3 items in wardrobe to generate profile')

            # Generate profile using AI
            profile = gemini_service.generate_style_profile(wardrobe_items)
            print(f'[DEBUG] AI profile: {profile}')

            # Add statistics
            stats = wardrobe_service.get_statistics(user_id)

            result = {
                'profile': profile,
                'statistics': stats,
                'itemCount': len(wardrobe_items),
                'generatedAt': datetime.now().isoformat()
            }

            return result
            
        except Exception as e:
            raise ValueError(f'Failed to generate profile: {str(e)}')
    
    def get_recommendations(self, item_id: int) -> Dict[str, Any]:
        """
        Get style recommendations based on a specific item
        
        Args:
            item_id: ID of the reference item
            
        Returns:
            Dict containing recommendations
        """
        try:
            # Get the item
            item = wardrobe_service.get_item_by_id(item_id)
            if not item:
                raise ValueError(f'Item {item_id} not found')
            
            # Get all wardrobe items
            wardrobe_items = wardrobe_service.get_all_items()
            
            if len(wardrobe_items) < 2:
                raise ValueError('Need at least 2 items in wardrobe for recommendations')
            
            # Get recommendations from AI
            recommendations = gemini_service.find_similar_items(item, wardrobe_items)
            
            result = {
                'referenceItem': item,
                'recommendations': recommendations,
                'generatedAt': datetime.now().isoformat()
            }
            
            return result
            
        except Exception as e:
            raise ValueError(f'Failed to get recommendations: {str(e)}')

# Create singleton instance
style_analysis_service = StyleAnalysisService()
